File: sdk/src/main/java/org/coralibre/android/sdk/internal/datatypes/ENInterval.java
package org.coralibre.android.sdk.internal.datatypes;

import org.coralibre.android.sdk.internal.datatypes.util.ENIntervalUtil;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.security.InvalidParameterException;
import java.util.Objects;

public class ENInterval {
    public static final long MAX_UINT_32 = 4294967295L;
    public static final int LONG_BYTES = 8;
    public static final int INT_BYTES = 4;


    private final long val;


    public ENInterval(final ENInterval enInterval) {
        this.val = enInterval.get();
    }

    public ENInterval(final long rawENNumber) {
        val = toEnIntervalNumberChecked(rawENNumber, false);
    }

    public ENInterval(final long input, final boolean isUnixTime) {
        val = toEnIntervalNumberChecked(input, isUnixTime);
    }

    public ENInterval(final byte[] rawENNumber) {
        if(rawENNumber.length != INT_BYTES)
            throw new InvalidParameterException("rawENNumber has wrong size");
        byte[] rawBuffer = new byte[LONG_BYTES];
        System.arraycopy(rawENNumber, 0, rawBuffer, 0, INT_BYTES);
        ByteBuffer buffer = ByteBuffer.wrap(rawBuffer);
        buffer.order(ByteOrder.LITTLE_ENDIAN);
        val = toEnIntervalNumberChecked(buffer.getLong(), false);
    }

    private long toEnIntervalNumberChecked(long input, boolean isUnixTime) {
        long result = input;
        if(isUnixTime)
            result = ENIntervalUtil.intervalNumberFromUnixTimestamp(input);

        if(result < 0 || MAX_UINT_32 < result)
            throw new InvalidParameterException("ENIntervalTimestamp out of bound: "
            + String.format("0x%X", result));
        return result;
    }

    public long get() {
        return val;
    }

    public long getUnixTime() {
        return ENIntervalUtil.intervalNumberToUnixTimestamp(val);
    }

    public byte[] getBytes() {
        // This has to output the value as Little endian
        ByteBuffer buffer = ByteBuffer.allocate(LONG_BYTES);
        buffer.order(ByteOrder.LITTLE_ENDIAN);
        buffer.putLong(val);
        byte[] retVal = new byte[INT_BYTES];
        System.arraycopy(buffer.array(), 0, retVal, 0, INT_BYTES);
        return retVal;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        ENInterval enInterval = (ENInterval) o;
        return val == enInterval.val;
    }

    @Override
    public int hashCode() {
        return Objects.hash(val);
    }

    @Override
    public String toString() {
        return Long.toString(val);
    }
}

File: sdk/src/main/java/org/coralibre/android/sdk/internal/datatypes/ENInterval.kt
package org.coralibre.android.sdk.internal.datatypes

import org.coralibre.android.sdk.internal.datatypes.util.ENIntervalUtil
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.security.InvalidParameterException
import java.util.Objects

class ENInterval {
    private val value: Long

    constructor(enInterval: ENInterval) {
        value = enInterval.get()
    }

    constructor(rawENNumber: Long) {
        value = toEnIntervalNumberChecked(rawENNumber, false)
    }

    constructor(input: Long, isUnixTime: Boolean) {
        value = toEnIntervalNumberChecked(input, isUnixTime)
    }

    constructor(rawENNumber: ByteArray) {
        if (rawENNumber.size != INT_BYTES) throw InvalidParameterException("rawENNumber has wrong size")
        val rawBuffer = ByteArray(LONG_BYTES)
        System.arraycopy(rawENNumber, 0, rawBuffer, 0, INT_BYTES)
        val buffer = ByteBuffer.wrap(rawBuffer)
        buffer.order(ByteOrder.LITTLE_ENDIAN)
        value = toEnIntervalNumberChecked(buffer.long, false)
    }

    private fun toEnIntervalNumberChecked(input: Long, isUnixTime: Boolean): Long {
        val result = if (isUnixTime) {
            ENIntervalUtil.intervalNumberFromUnixTimestamp(input)
        } else {
            input
        }

        if (result < 0 || MAX_UINT_32 < result) {
            throw InvalidParameterException(
                "ENIntervalTimestamp out of bound: " + String.format("0x%X", result)
            )
        }
        return result
    }

    fun get(): Long {
        return value
    }

    val unixTime: Long
        get() = ENIntervalUtil.intervalNumberToUnixTimestamp(value)

    // This has to output the value as Little endian
    val bytes: ByteArray
        get() {
            // This has to output the value as Little endian
            val buffer = ByteBuffer.allocate(LONG_BYTES)
            buffer.order(ByteOrder.LITTLE_ENDIAN)
            buffer.putLong(value)
            val retVal = ByteArray(INT_BYTES)
            System.arraycopy(buffer.array(), 0, retVal, 0, INT_BYTES)
            return retVal
        }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val enInterval = other as ENInterval
        return value == enInterval.value
    }

    override fun hashCode(): Int {
        return Objects.hash(value)
    }

    override fun toString(): String {
        return value.toString()
    }

    companion object {
        const val MAX_UINT_32 = 4294967295L
        const val LONG_BYTES = 8
        const val INT_BYTES = 4
    }
}
